### DISCUSSION TITLE
Allow responses from button or keyboard?

### DISCUSSION THREAD

**Discussion by AlecMRogers** [jspsych/jsPsych#1825] [Q&A] [answered]
I want to allow responses from both html buttons and the keyboard.

Should I try to add keyboard input to html-button-response, 
buttons to html-keyboard-response,
or is there some other existing option that allows both?

Thanks,
-alec

**Comment by becky-gilbert** [comment, marked as answer]
**EDIT**: Sorry, one last step that I forgot to mention. At the end of the trial you have to clear the keyboard listener so that it doesn't continue running during subsequent trials. You can do this in the `end_trial` function using the `jsPsych.pluginAPI.cancelAllKeyboardResponses()` function. Or if you assign `getKeyboardResponse` to a variable (e.g. `listener_id`) then you can clear that one specifically with `jsPsych.pluginAPI.cancelKeyboardResponse(listener_id)`.

---

Hi @AlecMRogers, sorry for the slow response! Not sure if this is still useful...

Unfortunately there's no other option (that I know of) to combine button and keyboard responses - I think you'll have to add one of the response options to a plugin that allows the other type of response. If it were me, I would probably add a keyboard listener to the button response plugin. 
1. In the `plugin.info` section, you'll want to add the key choices as an additional parameter. 
2. In the `plugin.trial` function, you can set up the keyboard listener with the [`jsPsych.pluginAPI.getKeyboardResponse` function](https://www.jspsych.org/core_library/jspsych-pluginAPI/#jspsychpluginapigetkeyboardresponse). Assuming you only need to capture the first valid response, it will look like this:
```js
jsPsych.pluginAPI.getKeyboardResponse({
  callback_function: after_response,
  valid_responses: trial.key_choices, 
  rt_method: 'performance',
  persist: false
});
``` 
Where `key_choices` is the name of your new key choices parameter, and `after_response` is the name of the function that should run after a valid key was pressed. The `after_response` function will be passed a single argument, an object containing the `key` and `rt` for the keyboard response ([example here](https://www.jspsych.org/core_library/jspsych-pluginAPI/#examples_2)). Inside this function, you should save the response to the trial data and do whatever else should happen next, e.g. end the trial.

I think that's it. Let us know if you've got it working, and feel free to follow up here if you need further guidance.

**Reply by AlecMRogers** [reply]
@becky-gilbert 
Thanks, I followed your advice, and the new plugin seems to work fine.

However, a jsPsych.pluginAPI.getKeyboardResponse is left in place, and my experiment errors out in a subsequent trial which uses a html-slider, probably because it does not remove the keyboard handler when a button is pressed.

Can you suggest the cleanest way to resolve that, and how can I send you a Starbucks gift card number?

**Reply by AlecMRogers** [reply]
Here is a solution that seems to work, in case others find this thread useful:

// store the listener when it is installed
var keyboardListener = jsPsych.pluginAPI.getKeyboardResponse ...

// remove keyboard handler in the end_trial() function
jsPsych.pluginAPI.cancelKeyboardResponse(keyboardListener);

**Reply by becky-gilbert** [reply]
Sorry for missing that last step, and glad you figured it out! I'll edit my answer to make it clearer for others.
